### ISSUE TITLE
The usage of ALL any ANY with the @auth directive

### ISSUE THREAD

**Issue by darrellwarde** [neo4j/graphql#293]
@darrellwarde Hey, a bit late to the party, but I think the same behaviour should apply to `@auth where & bind` in case we are referencing a relationship that is a list

With the following schema:
```graphql
type User {
  id: ID! @id
  name: String!
}
type Chat {
  id: ID! @id
  title: String @readonly
  description: String @readonly
  users: [User] @relationship(type: "HAS_USER", direction: OUT)
}

extend type Chat
  @auth(rules: [
    { operations: [CREATE, READ, UPDATE, DELETE], where: { users: { id: "$jwt.sub" } } }
  ])
```

How can I get all `Chats` for the logged user? Because `where` & `bind` apply `ALL()`, I will be able to get only chats where `$jwt.sub` is the only user.

On the other hand, I think `ALL()` makes sense if it was a singular relationship:
```graphql
type Chat {
  user: User @relationship(type: "HAS_USER", direction: OUT)
}
```

_Originally posted by @MarianAlexandruAlecu in https://github.com/neo4j/graphql/issues/190#issuecomment-873641640_

**Comment by MarianAlexandruAlecu** [comment]
@darrellwarde an even better solution would be allow devs to specify if they want `ALL` or `ANY`.

E.g.
```graphql
extend type Chat
  @auth(rules: [
    { operations: [CREATE, READ, UPDATE, DELETE], where: { users: { id: "$jwt.sub" } }, predicate: ANY }
  ])
```

Is this too cumbersome? What do you think? 🤔

**Comment by danstarns** [comment]
> ```graphql
> extend type Chat
>   @auth(rules: [
>     { operations: [CREATE, READ, UPDATE, DELETE], where: { users: { id: "$jwt.sub" } }, predicate: ANY }
>   ])
> ```

This is interesting, I don't think your suggested location is the best and something like this would be better IMO: 

```gql
extend type Chat
    @auth(rules: [{ operations: [CREATE, READ, UPDATE, DELETE], where: { ANY: { users: { id: "$jwt.sub" } } } }])
```

**Comment by MarianAlexandruAlecu** [comment]
@danstarns Your version is perfect! I'm looking forward to use it in my code 😺

**Comment by rcbevans** [comment]
@danstarns  Is this something being worked on?

I just ran into this exact issue today, and by hacking https://github.com/neo4j/graphql/blob/eb4d0c11a183f01b18f652a99319d6701c3e69f3/packages/graphql/src/translate/create-auth-and-params.ts#L120  to add a check to include `|| kind === "where"` I was able to get the where behaving as I had hoped it would.

I'm currently prototyping a Neo4j + GraphQL library backend and this issue is a show stopper (if we need to write resolvers to work around this then we lose all the amazing benefits GraphQL library offers).  My graph is several slightly overlapping subgraphs and so requires these ANY() where conditions to scope access down to portions of the subgraph to which the user should have access.

If this is something you don't have the cycles for, I'm totally willing to contribute and do the work to get it done.

**Comment by rcbevans** [comment]
> > ```graphql
> > extend type Chat
> >   @auth(rules: [
> >     { operations: [CREATE, READ, UPDATE, DELETE], where: { users: { id: "$jwt.sub" } }, predicate: ANY }
> >   ])
> > ```
> 
> This is interesting, I don't think your suggested location is the best and something like this would be better IMO:
> 
> ```graphql
> extend type Chat
>     @auth(rules: [{ operations: [CREATE, READ, UPDATE, DELETE], where: { ANY: { users: { id: "$jwt.sub" } } } }])
> ```

@danstarns I have a PR to add what I think you were describing here.  Let me know if you had something else in mind!

**Comment by rcbevans** [comment]
I had a PR up for this, but there's now a preference from the Neo4j core team to follow a similar design to the query `where` filters, using a postfix pattern like.

```
where: { users_INCLUDES: {id: "$jwt.sub" } }
```

This makes it easier to support a wider range of rules, either now or in future.

**Comment by rcbevans** [comment]
A new PR with the suggested api is available for review, #500

**Comment by darrellwarde** [comment]
Hey everyone! In 3.23.0, we released our new authorization features which will replace `@auth` in version 4.0.0, and this feature is now supported! 🎉  We would encourage you to [migrate](https://neo4j.com/docs/graphql-manual/current/guides/v4-migration/authorization/) as soon as possible so we can get your valuable feedback and you can enjoy all of the new features!

The new `@authorization` directive uses the same syntax as the filtering of the generated API, so all of the `_SOME`, `_ALL`, etc. predicates can now be used as desired.
